ddl/ddl_api.go

```diff
@@ -7077,6 +7150,24 @@ func BuildAddedPartitionInfo(ctx sessionctx.Context, meta *model.TableInfo, spec
 				return nil, ast.ErrPartitionsMustBeDefined.GenWithStackByArgs(meta.Partition.Type)
 			}
 		}
+	case model.PartitionTypeHash, model.PartitionTypeKey:
+		switch spec.Tp {
+		case ast.AlterTableCoalescePartitions:
+			if int(spec.Num) >= len(meta.Partition.Definitions) {
+				return nil, dbterror.ErrDropLastPartition
+			}
+			resetNum := meta.Partition.Num
+			defer func() { meta.Partition.Num = resetNum }()
+			meta.Partition.Num = uint64(len(meta.Partition.Definitions)) - spec.Num
```

Comment:
I think change `meta.Partition.Num` here is unsafe. Since the `meta` is read from memory infoschema, the `meta` should be read only, change it may affect another session.

Reply:
Refactored